csrc/fmhaReduction.cu

```diff
@@ -0,0 +1,357 @@
+    for (int ii = 0; ii < NumEltsPer16BVec / 2; ++ii) {
+      float2& f2 = reinterpret_cast<float2*>(outputVals)[ii];
+      mul(f2, f2, normalizedScale2);
+    }
+
+    // Convert the float values to DtypeO, and Store it to global memory.
+    if (isValidRow) {
+      convertAndStoreToGmem<DtypeO>(reinterpret_cast<char*>(oPtr + gmemStoreOffset), outputVals);
+    }
+  }
+
+// Trigger the secondary kernel.
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  cudaTriggerProgrammaticLaunchCompletion();
+#endif
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#define SELECT_FMHA_REDUCTION_KERNEL(HeadDimPerCta)                                               \
+  if (kernelMeta.mDataTypeQ == DATA_TYPE_E4M3) {                                                  \
+    if (kernelMeta.mDataTypeO == DATA_TYPE_E4M3) {                                                \
+      kernel = &fmhaReductionKernel<64, 512, HeadDimPerCta, true, __nv_fp8_e4m3, half>;           \
+    } else if (kernelMeta.mDataTypeO == DATA_TYPE_FP16) {                                         \
+      kernel = &fmhaReductionKernel<64, 512, HeadDimPerCta, true, half, half>;                    \
+    } else if (kernelMeta.mDataTypeO == DATA_TYPE_BF16) {                                         \
+      kernel = &fmhaReductionKernel<64, 512, HeadDimPerCta, true, __nv_bfloat16, __nv_bfloat16>;  \
+    } else {                                                                                      \
+      TORCH_CHECK(false, "Not implemented");                                                      \
+    }                                                                                             \
+  } else {                                                                                        \
+    TORCH_CHECK(kernelMeta.mDataTypeQ == kernelMeta.mDataTypeO, "Not implemented");               \
+    if (kernelMeta.mDataTypeQ == DATA_TYPE_FP16) {                                                \
+      kernel = &fmhaReductionKernel<64, 512, HeadDimPerCta, false, half, half>;                   \
+    } else if (kernelMeta.mDataTypeQ == DATA_TYPE_BF16) {                                         \
+      kernel = &fmhaReductionKernel<64, 512, HeadDimPerCta, false, __nv_bfloat16, __nv_bfloat16>; \
+    } else {                                                                                      \
+      TORCH_CHECK(false, "Not implemented");                                                      \
+    }                                                                                             \
+  }
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void runFmhaReduction(TllmGenFmhaKernelMetaInfo const& kernelMeta, KernelParams const& params,
+                      int32_t multiProcessorCount, cudaStream_t stream, bool enable_pdl) {
```

Comment:
a minor request is to reorder `stream` and `enable_pdl` to be consistent of the remaining codebase
```suggestion
                      int32_t multiProcessorCount, bool enable_pdl, cudaStream_t stream) {
```